Add proper descriptions to some `settings` option types

rnhmjoj opened the pull request:
###### Motivation for this change
Add proper descriptions to some `settings` option types. See https://github.com/NixOS/nixpkgs/issues/86402#issuecomment-624516647

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change
- [x] Built the NixOS manual
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rnhmjoj committed:
nixos/clight: add type description to the `settings` option

rnhmjoj committed:
nixos/pdns-recursor: add type description to the `settings` option

bjornfor commented on lib/types.nix:
```diff
@@ -583,6 +583,9 @@ rec {
       functor = (defaultFunctor name) // { wrapped = [ t1 t2 ]; };
     };
 
+    oneOrMore = t: either t (listOf t)
```
Data point: I expected oneOrMore to be 'listOf t' with minimum list length of 1, not two "separate" types.

rnhmjoj replied:
Well, you are right but in the context of a subset of Nix mapping to INI or TOML etc. it makes a difference. For example, if the top-level of an attrset is a `either t or (listOf t)`, this would be valid:
```nix
{ 
  database.address = "127.0.0.1";
  database.ports = [ 8001 ]; 
}
```
and would turn into something like
```TOML
[database]
address = "127.0.0.1"
ports = [ 8001 ]
```
On the other hand, `database.address = "127.0.0.1"` would not typecheck because it's not a list.

Maybe the name `oneOrMore` is not appropriate, because it appears to be a haskell-like `NonEmpty` but it's actually a different sum-type.

jtojnar replied:
Yeah, I would not expect a type called `oneOrMore` to allow empty list as a value.

As is, the type is just a regular `listOf` which accepts omitting two characters (`[` and `]`) when passing a singleton. I do not see much value in that.

rnhmjoj replied:
Agreed, it's not really that useful.

rnhmjoj committed:
lib/types: improve type descriptions

This commit tries to improve the readability of the Nixpkgs
types descriptions by adding parenthesis when needed.
In this way the nested structures made when composing types
is preserved in the description. For example, this expression

    attrsOf (attrsOf (oneOf [ bool int str (listOf str) ]))

will produce the following description:

  attribute set of (attribute set of (one of
  [ boolean, signed integer, string, (list of string) ]))

rnhmjoj committed:
nixos/gerrit: improve settings type description

rnhmjoj committed:
nixos/smartdns: improve settings type description

rnhmjoj commented:
After implementing the parenthesised descriptions I found a couple more offenders.
On second thought, a first class `oneOrMore` type could still be useful if we provide a better description than the defeault. If the subtype is complex enough, I think
```
Type: one or a list of (one of [ string, signed integer, boolean ])
```
is better than the default
```
either (one of [ string, signed integer, boolean ]) or (list of (one of [ string, signed integer, boolean ]))
```

rnhmjoj commented:
@Infinisil I don't know if you remember, a few months ago on IRC we chatted about improving type descriptions and this is basically my attempt at implementing the idea you suggested. Can you tell me what you think?

SuperSandro2000 commented:
@ofborg eval

Infinisil commented on lib/types.nix:
```diff
@@ -17,6 +17,18 @@ rec {
+  # enclosing expression with precedence `d`. Since Nix does
+  # not have arbitrary infix operators, only two precedences
+  # are defined: 1 for function application, which may need
+  # to be parenthesised, and 0 for constants.
```
We do have an infix operator for the types though: `either`. With this PR, `either int (either bool str)` shows as
```
signed integer or (boolean or string)
```

Which I think isn't correct, because `either` is left and right-associative, so it should be
```
signed integer or boolean or string
```

This would then also obviate the need for the `oneOf` change.

And I believe to implement this, we need a `descriptionPrec = d: ...` field on option types. This then also mirrors Haskell's `showsPrec` more closely.

rnhmjoj commented:
Closing as https://github.com/NixOS/nixpkgs/pull/191353 implemented an alternative solution.
Honestly, I think this appoarch was better, but it's too late.
